paimon-spark/paimon-spark-common/src/main/java/org/apache/paimon/spark/catalog/functions/BucketFunction.java

```diff
@@ -0,0 +1,724 @@
+    }
+
+    @Override
+    public String name() {
+        return "bucket";
+    }
+
+    /** Bound bucket function for generic type. */
+    public static class BucketGeneric implements ScalarFunction<Integer> {
+        protected final DataType[] bucketKeyTypes;
+        protected final BinaryRow bucketKeyRow;
+        // not serializable
+        protected transient BinaryRowWriter bucketKeyWriter;
+        private transient ValueWriter[] valueWriters;
+
+        public BucketGeneric(DataType[] sqlTypes) {
+            this.bucketKeyTypes = sqlTypes;
+            this.bucketKeyRow = new BinaryRow(bucketKeyTypes.length);
+            this.bucketKeyWriter = new BinaryRowWriter(bucketKeyRow);
+            this.valueWriters = createValueWriter(bucketKeyTypes);
+        }
+
+        private void readObject(ObjectInputStream in) throws IOException, ClassNotFoundException {
+            in.defaultReadObject();
+            this.bucketKeyWriter = new BinaryRowWriter(bucketKeyRow);
+            this.valueWriters = createValueWriter(bucketKeyTypes);
+        }
+
+        private static ValueWriter[] createValueWriter(DataType[] columnTypes) {
+            ValueWriter[] writers = new ValueWriter[columnTypes.length];
+            for (int i = 0; i < columnTypes.length; i += 1) {
+                writers[i] = ValueWriter.of(columnTypes[i]);
+            }
+
+            return writers;
+        }
+
+        @Override
+        public Integer produceResult(InternalRow input) {
+            bucketKeyWriter.reset();
+            for (int i = 0; i < valueWriters.length; i += 1) {
+                valueWriters[i].write(bucketKeyWriter, i, input, i + 1);
+            }
+
+            bucketKeyWriter.complete();
+            return bucket(bucketKeyRow, input.getInt(NUM_BUCKETS_ORDINAL));
+        }
+
+        @Override
+        public DataType[] inputTypes() {
+            DataType[] inputTypes = new DataType[bucketKeyTypes.length + 1];
+            inputTypes[0] = DataTypes.IntegerType;
+            for (int i = 0; i < bucketKeyTypes.length; i += 1) {
+                inputTypes[i + 1] = bucketKeyTypes[i];
+            }
+
+            return inputTypes;
+        }
+
+        @Override
+        public DataType resultType() {
+            return DataTypes.IntegerType;
+        }
+
+        @Override
+        public boolean isResultNullable() {
+            return false;
+        }
+
+        @Override
+        public String name() {
+            return "bucket";
+        }
+
+        @Override
+        public String canonicalName() {
+            return String.format(
+                    "paimon.bucket(%s)",
+                    Arrays.stream(bucketKeyTypes)
+                            .map(DataType::catalogString)
+                            .collect(Collectors.joining(", ")));
+        }
+
+        // org.apache.paimon.table.sink.KeyAndBucketExtractor.bucket(int, int)
+        protected static int bucket(BinaryRow bucketKey, int numBuckets) {
```

Comment:
Why not direct use `KeyAndBucketExtractor.bucket`

Reply:
I try to introduce the HashType for the paimon table https://github.com/apache/paimon/pull/5445. So I'd like to see the hash calculation are managed in one place

Reply:
I'm trying to avoid the uncessary `checkArgument`  validate, I can use that directly as the overhead is small

Reply:
Oh,, I alse noticed this check and planned to delete it